Automattic/wp-calypso
# Refactor TeamCity templates

This is an *experimental* PR. We talked about TeamCity templates vs Kotlin classes ages ago, and I just wanted to see how it would look like.

## Background

We use TeamCity's [Build Configuration Templates](https://www.jetbrains.com/help/teamcity/build-configuration-template.html) to abstract some common parts of building WPCom plugins in TeamCity.

While this works ok, it's feels a bit clunky when we use Kotlin DSL, as we can use just regular programming patterns (functions, classes...) to achieve simialr abstractions.

We introduced Templates in WPCom builds when we didn't have lots of expertise in Kotlin. Since then, other builds (in particular, e2e buids) have use Kotlin featues to abstract common parts.

## Changes

Refactor our build config to avoid using Templates, and use a regular Kotlin class instead.

### Pros

* As it is "100% Kotlin" based, so to speak, it is easier to get IDE features like autocomplete, and maybe even test if we want to.

### Cons

* It feels like our own DSL on top of the existing DSL. That usually ends up being a lot of maintenance and something new to learn for people that know the original DSL. But maybe is because I don't know enough Kotlin to make it more idiomatic.


## Testing instructions

* Verify all builds are green *and* there are no errors applying the build settings.
* Inspect the config of WPCom builds and verify they seem correct.
